### PR TITLE
tasks: add tablet migration virtual task

### PR DESCRIPTION
**Pull request by Deexie**
In this change, tablet_virtual_task starts supporting tablet
migration, in addition to tablet repair. Both tablet operations 
reuse the same virtual_task because their task data is retrieved
similarly. However, it changes nothing from the task manager 
API users' perspective. They can list running migrations or check 
their statuses all the same as if migration had its own virtual_task. 

Users can see running migration tasks - finished tasks are not 
presented with the task manager API. However, the result 
of the migration (whether it succeeded or failed) would be
presented to users, if they use wait API.

If a migration was reverted, it will appear to users as failed.
We assume that the migration was reverted, when its destination
does not contain a tablet replica.

Fixes: https://github.com/scylladb/scylladb/issues/21365.

No backport, new feature

### CODE REVIEW

**Comment by Deexie** [comment]
Draft until https://github.com/scylladb/scylladb/pull/21624 is merged

**Comment by Deexie** [comment]
- rebase
- split the commits
- change the way state of a task is retrieved in wait

**Comment by Deexie** [comment]
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13567/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13567/) **The following new/updated tests ran 100 times for each mode:** 🔹 topology_tasks/test_tablet_tasks ❌ - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/486/) ❌ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2012/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5122/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13137/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13567/)
> #### Failed Tests (2/36827):
> 
>     * [test_slow_query_logging](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13567/testReport/junit/alternator_tests/TesterAlternator/Tests___dtest_with_tablets___test_slow_query_logging) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_slow_query_logging)
> 

https://github.com/scylladb/scylladb/issues/21490

>     * [cqlpy.cassandra_tests/validation/entities/frozen_collections_test.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13567/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_frozen_collections_test_debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/frozen_collections_test.debug.2)
> 

https://github.com/scylladb/scylladb/issues/21292

> 
> #### Build Details:
> 
>     * Duration: 8 hr 20 min
> 
>     * Builder: i-0fe768f14eaaef9f5 (m5ad.8xlarge)

**Comment by Deexie** [comment]
- fix getting status

**Comment by Deexie** [comment]
- add test for a failed migration task

File: replica/tablets.cc [hunk 1]
```diff
@@ -73,6 +74,7 @@ schema_ptr make_tablets_schema() {
             .with_column("repair_time", timestamp_type)
             .with_column("repair_task_info", tablet_task_info_type)
             .with_column("repair_scheduler_config", repair_scheduler_config_type, column_kind::static_column)
+            .with_column("migration_task_info", tablet_task_info_type)
```
Comment (denesb) [inline replica/tablets.cc:77]: Please extend the save/load tests in boost/test_tablets.cc with this new column as well, to ensure we save and load it.

File: service/task_manager_module.cc [hunk 2]
```diff
@@ -154,6 +154,20 @@ table_id tablet_virtual_task::get_table_id(const tasks::virtual_task_hint& hint)
...
 }
 
+locator::tablet_task_type tablet_virtual_task::get_task_type(const tasks::virtual_task_hint& hint) const {
+    if (!hint.task_type.has_value()) {
+        on_internal_error(tasks::tmlogger, "tablet_virtual_task hint does not contain task type");
+    }
+    return hint.task_type.value();
+}
+
+locator::tablet_id tablet_virtual_task::get_tablet_id(const tasks::virtual_task_hint& hint) const {
```
Comment (denesb) [inline service/task_manager_module.cc:230]: Shouldn't these be methods on the task hint object itself?

File: service/task_manager_module.cc [hunk 3]
```diff
@@ -71,11 +71,8 @@ future<std::optional<tasks::virtual_task_hint>> tablet_virtual_task::contains(ta
 }
 
 future<tasks::is_abortable> tablet_virtual_task::is_abortable(tasks::virtual_task_hint hint) const {
-    auto task_type = get_task_type(hint);
-    if (!is_repair_task(task_type)) {
-        return make_ready_future<tasks::is_abortable>(tasks::is_abortable::no);
-    }
-    return make_ready_future<tasks::is_abortable>(tasks::is_abortable::yes);
+    auto task_type = get_task_type(std::move(hint));
+    return make_ready_future<tasks::is_abortable>(is_repair_task(task_type));
```
[the remark this answers is not in the document]
Comment (Deexie) [reply service/task_manager_module.cc:75]: The previous implementation returned the default value for migration tasks, while the current implementation returns the expected is_abortable value. The result is the same, but now the decision is conscious

[the remark this answers is not in the document]
Comment (denesb) [reply service/task_manager_module.cc:75]: I see, fair enough.

**Comment by Deexie** [comment]
- hint getters are now methods of virtual_task_hint
- mention in a commit description that the tablet_virtual_task methods will start supporting migration later
- add migration_task_info check to tablet_test.cc

**Comment by Deexie** [comment]
- rebase

File: tasks/virtual_task_hint.cc [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+locator::tablet_id virtual_task_hint::get_tablet_id() const {
+    if (!tablet_id.has_value()) {
+        on_internal_error(tasks::tmlogger, "tablet_virtual_task hint does not contain tablet_id");
+    }
+    return tablet_id.value();
+}
+
+}
```
Comment (denesb) [inline tasks/virtual_task_hint.cc:35]: Please avoid tiny `.cc` files like this, they have significant linking overhead compared to their size. you can just put these method implementations in `task_manager.cc`.

File: test/boost/tablets_test.cc [hunk 5]
```diff
@@ -152,7 +158,10 @@ SEASTAR_TEST_CASE(test_tablet_metadata_persistence) {
                 tmap.set_tablet(tb, tablet_info {
                     tablet_replica_set {
                         tablet_replica {h2, 2},
-                    }
+                    },
+                    {},
+                    {},
```
Comment (denesb) [inline test/boost/tablets_test.cc:163]: At least one test should put something interesting here.

**Comment by Deexie** [comment]
- move virtual_task_hint.cc content to task_manager.cc
- add repair related fields to tablet_map_to_mutation and check them in a test

**Comment by Deexie** [comment]
- set repair time in  tablet_map_to_mutation only if it does not contain default value

**Pull request merged**